migrations/blocked_tasks.js

```diff
@@ -0,0 +1,74 @@
+function taskIsBlocked(task) {
+    for(dep of task.depends_on) {
+        if (dep.unattainable) {
+            return true
+        }
+    }
+
+    return false
+}
+
+function unattainableResolved(task) {
+    for (dep of task.depends_on) {
+        if (dep.unattainable == null) {
+            return false
+        }
+    }
+    return true
+}
+
+finishedStatuses = ["success","failed"]
+millisecondsPerDay = 86400000
+delta = millisecondsPerDay * 1
+
+startingDate = ISODate("2016-01-01T00:00:00Z")
+while(startingDate < new Date()) {
+    loops = 0
+    tasksSize = 0
+    while (true) {
+        tasks = db.tasks.find({"create_time": {"$gte": startingDate, "$lte": new Date(startingDate.getTime() + delta)}, "status":"undispatched", "depends_on.status": {"$in": ["success", "failed", "", "*"]}, "depends_on": {"$elemMatch":{"unattainable": {"$exists": false}}}}, {"depends_on":1}).toArray()
+        printjson(loops++)
+        if (tasks.length == 0 || tasks.length == tasksSize) {
+            break
+        }
+        tasksSize = tasks.length
+
+        for (i=0; i < tasks.length; i++) {
+            taskUpdated = false
+            dependsOn = tasks[i].depends_on
+            for (j=0; j < dependsOn.length; j++) {
+                if(dependsOn[j].unattainable != null) {
+                    continue
+                }
+                
+                depTask = db.tasks.findOne({"_id": dependsOn[j]._id}, {"status": 1, "depends_on": 1});
+                if(depTask == null) {
+                    printjson(tasks[i])
+                    break
+                }
+
+                if (finishedStatuses.includes(depTask.status)) {
+                    taskUpdated = true
+                    // 1st degree blocked
+                    if(depTask.status != "*" && depTask.status != dependsOn[j].status) {
```

Comment:
I don't think the status of a task can be `*`. `depTask` is a task from the tasks collection. Did you mean to check `dependsOn`?

Reply:
fixed 😰